src/proxy/topic/ReliableTopicProxy.ts

```diff
@@ -204,4 +268,22 @@ export class ReliableTopicProxy<E> extends BaseProxy implements ITopic<E> {
         }).catch(deferred.reject);
     }
 
+    private addMessagesOrFail(messages: ReliableTopicMessage[]): Promise<void> {
+        return this.ringbuffer.addAll(messages, OverflowPolicy.FAIL).then((sequenceId: Long) => {
+            if (sequenceId.toNumber() === -1) {
+                throw new TopicOverloadError(`Failed to publish messages on topic ${this.name}.`);
+            }
+        });
+    }
+
+    private addAndBlock(messages: ReliableTopicMessage[], pauseMillis: number): Promise<void> {
+        this.ringbuffer.addAll(messages, OverflowPolicy.FAIL).then((id: Long) => {
+            if (id.toNumber() === -1) {
+                setTimeout(() => {
+                    this.addAndBlock(messages, Math.min(pauseMillis * 2, 2000));
+                }, pauseMillis);
+            }
+        });
+        return Promise.resolve();
```

Comment:
```suggestion
   private addAndBlock(returnPromise: DeferredPromise<void>, messages: ReliableTopicMessage[], pauseMillis: number): void {
        this.ringbuffer.addAll(messages, OverflowPolicy.FAIL).then((id: Long) => {
            if (id.toNumber() === -1) {
                setTimeout(() => {
                    this.addAndBlock(returnPromise, messages, Math.min(pauseMillis * 2, 2000));
                }, pauseMillis);
            } else {
                returnPromise.resolve();
            }
        }).catch(returnPromise.reject);
    }
```


Then in the publishAll: 

```ts
publishAll(messages: any[]): Promise<void> {
        const capacity = messages.length;
        const reliableTopicMessages: Array<ReliableTopicMessage> = Array<ReliableTopicMessage>(capacity);

        assertNotNull(messages);
        for (const message of messages) {
            assertNotNull(message);
        }

        const deferred = deferredPromise<void>();

        try {
            for (const message of messages) {
                const reliableTopicMessage = new ReliableTopicMessage();
                reliableTopicMessage.payload = this.serializationService.toData(message);
                reliableTopicMessage.publishTime = Long.fromNumber(Date.now());
                reliableTopicMessages.push(reliableTopicMessage);
            }
            switch (this.overloadPolicy) {
                case TopicOverloadPolicy.ERROR:
                    return this.addMessagesOrFail(reliableTopicMessages);
                case TopicOverloadPolicy.DISCARD_NEWEST:
                    return this.ringbuffer.addAll(reliableTopicMessages, OverflowPolicy.OVERWRITE).then(() => {});
                case TopicOverloadPolicy.DISCARD_OLDEST:
                    return this.ringbuffer.addAll(reliableTopicMessages, OverflowPolicy.FAIL).then(() => {});
                case TopicOverloadPolicy.BLOCK:
                    this.addAndBlock(deferred, reliableTopicMessages, 100);
                    break;
                default:
                    return Promise.reject(new IllegalArgumentError('Unknown overload policy'));
            }
        } catch (e) {
            if (e instanceof SchemaNotReplicatedError) {
                return this.registerSchema(e.schema, e.clazz).then(() => this.publishAll(messages));
            }
            return Promise.reject(new HazelcastError(e + 'Failed to publish these messages' + messages + 'to topic' + this.name));
        }
        return deferred.promise;
    }
```

Reply:
I have to improve myself about promises.